File: Group_the_People_Given_the_Group_Size They_Belong_To.cpp
class Solution {
public:
    vector<vector<int>> groupThePeople(vector<int>& groupSizes) {
        vector<vector<int>> res;
        unordered_map<int, vector<int>> groups;

        for (int i = 0; i < groupSizes.size(); i++) {
            int size = groupSizes[i];

            groups[size].push_back(i);

            if (groups[size].size() == size) {
                res.push_back(groups[size]);
                groups.erase(size);
            }
        }

        return res;        
    }
};

File: Group_the_People_Given_the_Group_Size They_Belong_To.java
class Solution {
    public List<List<Integer>> groupThePeople(int[] groupSizes) {
        List<List<Integer>> res = new ArrayList<>();
        Map<Integer, List<Integer>> groups = new HashMap<>();

        for (int i = 0; i < groupSizes.length; i++) {
            int size = groupSizes[i];

            groups.putIfAbsent(size, new ArrayList<>());
            groups.get(size).add(i);

            if (groups.get(size).size() == size) {
                res.add(groups.get(size));
                groups.remove(size);
            }
        }

        return res;        
    }
}
